pkg/apiserver/registry/core/controllerregistration/strategy.go

```diff
@@ -42,6 +44,34 @@ func (controllerRegistrationStrategy) PrepareForUpdate(_ context.Context, obj, o
 	if mustIncreaseGeneration(oldControllerRegistration, newControllerRegistration) {
 		newControllerRegistration.Generation = oldControllerRegistration.Generation + 1
 	}
+
+	handleAutoEnabledResources(oldControllerRegistration, newControllerRegistration)
+}
+
+func handleAutoEnabledResources(oldControllerRegistration, newControllerRegistration *core.ControllerRegistration) {
+	resourceKindTypeToResource := map[string]core.ControllerResource{}
+	for _, resource := range oldControllerRegistration.Spec.Resources {
+		resourceKindTypeToResource[gardenerutils.ExtensionsID(resource.Kind, resource.Type)] = resource
+	}
+
+	for i, resource := range newControllerRegistration.Spec.Resources {
+		var (
+			oldResource core.ControllerResource
+			ok          bool
+		)
+
+		if oldResource, ok = resourceKindTypeToResource[gardenerutils.ExtensionsID(resource.Kind, resource.Type)]; !ok {
+			continue
+		}
+
+		// When globallyEnabled was set from true to false, the shoot type must be removed from the autoEnable list.
+		// Don't do anything if globallyEnabled was not set before, as this means a conflict and field AutoEnable should take precedence in this case.
+		if oldResource.GloballyEnabled != nil && *oldResource.GloballyEnabled && resource.GloballyEnabled != nil && !*resource.GloballyEnabled {
```

Comment:
I think defaulting happens before `PrepareForUpdate`

If the old resource had `.autoEnable=[shoot]` and `.globallyEnable=true` (or even if it did not have `.globallyEnabled` defined) and the new resource has `.autoEnable=[shoot]` and `.globallyEnabled=false`, it would first be defaulted in  https://github.com/gardener/gardener/blob/fb186fb4f0086cd926220840270a6d08ab7ae157/pkg/apis/core/v1beta1/defaults_controllerregistration.go#L27-L28 to `.globallyEnabled=true` before `PrepareForUpdate` is called, which leads to this function never getting executed.

Reply:
Thanks for the catch 👍 I improved the logic in 99a9904364a90386d76ab1abe808793da7eb7cbe.